v2/robotmk/src/bin/scheduler/environment.rs

```diff
@@ -1,122 +1,117 @@
-            let environment_build_status = run_environment_build(
-                ChildProcessSupervisor {
-                    command_spec: &build_instructions.command_spec,
-                    stdio_paths: Some(StdioPaths {
-                        stdout: stdio_directory.join(format!("{}.stdout", suite.id)),
-                        stderr: stdio_directory.join(format!("{}.stderr", suite.id)),
-                    }),
-                    timeout: build_instructions.timeout,
-                    cancellation_token: &suite.cancellation_token,
-                },
-                start_time,
-            )?;
-            let drop_suite = matches!(environment_build_status, EnvironmentBuildStatus::Failure(_));
-            environment_build_states_administrator.update(&suite.id, environment_build_status)?;
-            (!drop_suite).then_some(suite)
-        }
-        None => {
-            debug!("Nothing to do for suite {}", suite.id);
-            environment_build_states_administrator
-                .update(&suite.id, EnvironmentBuildStatus::NotNeeded)?;
-            Some(suite)
-        }
+) -> Result<BuildOutcome> {
+    let Some(instructions) = instructions else {
+        let outcome = BuildOutcome::NotNeeded;
+        debug!("Nothing to do for suite {}", id);
+        build_stage_reporter.update(id, EnvironmentBuildStage::Complete(outcome.clone()))?;
+        return Ok(outcome);
     };
-    Ok(suite)
+    info!("Building environment for suite {}", id);
+    let start_time = Utc::now();
+    build_stage_reporter.update(
+        id,
+        EnvironmentBuildStage::InProgress(start_time.timestamp()),
+    )?;
+    let outcome = run_environment_build(
+        ChildProcessSupervisor {
+            command_spec: &instructions.command_spec,
+            stdio_paths: Some(StdioPaths {
+                stdout: stdio_directory.join(format!("{}.stdout", id)),
+                stderr: stdio_directory.join(format!("{}.stderr", id)),
+            }),
+            timeout: instructions.timeout,
+            cancellation_token,
+        },
+        start_time,
+    )?;
+    build_stage_reporter.update(id, EnvironmentBuildStage::Complete(outcome.clone()))?;
+    Ok(outcome)
 }
 
 fn run_environment_build(
     build_process_supervisor: ChildProcessSupervisor,
     reference_timestamp_for_duration: DateTime<Utc>,
-) -> Result<EnvironmentBuildStatus> {
-    let child_process_outcome = match build_process_supervisor
-        .run()
-        .context("Environment building failed, suite will be dropped")
-        .map_err(log_and_return_error)
-    {
-        Ok(child_process_outcome) => child_process_outcome,
-        Err(error) => {
-            return Ok(EnvironmentBuildStatus::Failure(
-                EnvironmentBuildStatusError::Error(format!("{error:?}")),
-            ))
+) -> Result<BuildOutcome> {
+    let build_result = build_process_supervisor.run();
+    let child_process_outcome = match build_result {
```

Comment:
Maybe we could use `let ... else ...` here, but not sure.

Reply:
Let's do this separately after merging

Reply:
I was interested and took a look, I think `let ... else ...` cannot be applied here for precisely the reason you mentioned (you don't get the error variant in the `else` branch).